moveos/moveos-store/src/event_store/mod.rs

```diff
@@ -47,33 +58,77 @@ pub trait EventStore {
 }
 
 impl EventDBStore {
     pub fn new(instance: StoreInstance) -> Self {
         EventDBStore {
             event_store: EventDBBaseStore::new(instance.clone()),
-            indexer_store: EventIndexDBStore::new(instance),
+            indexer_store: EventIndexDBStore::new(instance.clone()),
+            event_handle_store: EventHandleDBStore::new(instance),
+        }
+    }
+
+    fn get_event_handle(&self, event_handle_id: ObjectID) -> Result<Option<EventHandle>> {
+        self.event_handle_store.kv_get(event_handle_id)
+    }
+
+    fn get_or_create_event_handle(&self, event_handle_type: StructTag) -> Result<EventHandle> {
+        let event_handle_id = EventHandle::derive_event_handle_id(event_handle_type);
+        let event_handle = self.get_event_handle(event_handle_id)?;
+        if let Some(event_handle) = event_handle {
+            return Ok(event_handle);
         }
+        let event_handle = EventHandle::new(event_handle_id, 0);
+        self.save_event_handle(event_handle.clone())?;
+        Ok(event_handle)
     }
 
-    pub fn save_event(&self, event: Event) -> Result<()> {
-        let key = (event.event_id.event_handle_id, event.event_id.event_seq);
-        self.event_store.kv_put(key, event)
+    fn save_event_handle(&self, event_handle: EventHandle) -> Result<()> {
+        self.event_handle_store
+            .put_all(vec![(event_handle.id, event_handle)])
     }
 
-    pub fn save_events(&self, events: Vec<Event>) -> Result<()> {
-        self.event_store.put_all(
-            events
-                .into_iter()
-                .map(|event| {
-                    (
-                        (event.event_id.event_handle_id, event.event_id.event_seq),
-                        event,
-                    )
-                })
-                .collect(),
-        )
+    pub fn save_events(&self, tx_events: Vec<TransactionEvent>) -> Result<Vec<EventID>> {
+        let event_types = tx_events
+            .iter()
+            .map(|event| event.event_type.clone())
+            .collect::<HashSet<_>>();
+        let mut event_handles = event_types
+            .into_iter()
+            .map(|event_type| {
+                let handle = self.get_or_create_event_handle(event_type.clone())?;
+                Ok((event_type, handle))
+            })
+            .collect::<Result<HashMap<_, _>>>()?;
+        let mut event_ids = vec![];
+        let events = tx_events
+            .into_iter()
+            .map(|tx_event| {
+                let handle = event_handles
+                    .get_mut(&tx_event.event_type)
+                    .expect("Event handle must exist");
+                let event_id = EventID::new(handle.id, handle.count);
+                let event = Event::new(
+                    event_id,
+                    tx_event.event_type,
+                    tx_event.event_data,
+                    tx_event.event_index,
+                );
+                handle.count += 1;
```

Comment:
If parallel optimization is done in the future, what impact will this have on parallelization and transaction rearrangement?

Reply:
I've thought about this and have come to a conclusion: parallel executing does not affect events. 

Because in Rooch, the transaction is ordered by the sequencer first, the parallel executing does change the order, so we need to ensure the events are written to store as the transaction order. The events stored in different nodes will guarantee consistency.